src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.SymbolMapping;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    [Export(typeof(IWpfTextViewConnectionListener))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Interactive)]
+    internal class DefinitionContextConnectionListener : IWpfTextViewConnectionListener
+    {
+        private readonly Dictionary<ITextView, CancellationTokenSource> _views = new Dictionary<ITextView, CancellationTokenSource>();
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+        private readonly ICodeDefinitionWindowService _codeDefinitionWindowService;
+
+        [ImportingConstructor]
+        public DefinitionContextConnectionListener(
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            ICodeDefinitionWindowService codeDefinitionWindowService)
+        {
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+            _codeDefinitionWindowService = codeDefinitionWindowService;
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (!_views.ContainsKey(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _views.Add(textView, new CancellationTokenSource());
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
+                var fireAndForget = UpdateDefinitionContext(textView, textView.Caret.Position);
+            }
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
```

Comment:
There shouldn't be.  This condition is supposed to be: either the view is closed, or the projection buffer has changed so that it no longer contains any source spans that come from Roslyn documents.